flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1262,11 +1261,19 @@ public void stopCheckpointScheduler() {
 	 */
 	public void abortPendingCheckpoints(CheckpointException exception) {
 		synchronized (lock) {
-			for (PendingCheckpoint p : pendingCheckpoints.values()) {
-				failPendingCheckpoint(p, exception.getCheckpointFailureReason());
-			}
+			// do not traverse pendingCheckpoints directly, because it might be changed
```

Comment:
? Is this a bug fix?

Reply:
No, it's a reminder. It's introduced by this PR.
After the refactoring of error handling, the `failPendingCheckpoint` is the only one entrance to fail a pending checkpoint. The `failPendingCheckpoint` maintains the `pendingCheckpoints`. We don't need to maintain `pendingCheckpoints` everywhere. However as side-effect, we can't invoke `failPendingCheckpoint` in a for loop like before. A `ConcurrentModificationException` will be raised in that way.